src/libValidator/Validator.cpp

```diff
@@ -148,3 +151,199 @@ bool Validator::CheckCreatedTransactionFromLookup(const Transaction& tx) {
+                                      const Container& commKeys) {
+  LOG_MARKER();
+
+  unsigned int index = 0;
+  unsigned int count = 0;
+
+  const vector<bool>& B2 = block.GetB2();
+  if (commKeys.size() != B2.size()) {
+    LOG_GENERAL(WARNING, "Mismatch: committee size = "
+                             << commKeys.size()
+                             << ", co-sig bitmap size = " << B2.size());
+    return false;
+  }
+
+  // Generate the aggregated key
+  vector<PubKey> keys;
+  for (auto const& kv : commKeys) {
+    if (B2.at(index)) {
+      keys.emplace_back(get<PubKey>(kv));
+      count++;
+    }
+    index++;
+  }
+
+  if (count != ConsensusCommon::NumForConsensus(B2.size())) {
+    LOG_GENERAL(WARNING, "Cosig was not generated by enough nodes");
+    return false;
+  }
+
+  shared_ptr<PubKey> aggregatedKey = MultiSig::AggregatePubKeys(keys);
+  if (aggregatedKey == nullptr) {
+    LOG_GENERAL(WARNING, "Aggregated key generation failed");
+    return false;
+  }
+
+  // Verify the collective signature
+  vector<unsigned char> serializedHeader;
+  block.GetHeader().Serialize(serializedHeader, 0);
+  block.GetCS1().Serialize(serializedHeader, serializedHeader.size());
+  BitVector::SetBitVector(serializedHeader, serializedHeader.size(),
+                          block.GetB1());
+  if (!Schnorr::GetInstance().Verify(serializedHeader, 0,
+                                     serializedHeader.size(), block.GetCS2(),
+                                     *aggregatedKey)) {
+    LOG_GENERAL(WARNING, "Cosig verification failed");
+    for (auto& kv : keys) {
+      LOG_GENERAL(WARNING, kv);
+    }
+    return false;
+  }
+
+  return true;
+}
+
+bool Validator::CheckDirBlocks(
+    const vector<boost::variant<DSBlock, VCBlock,
+                                FallbackBlockWShardingStructure>>& dirBlocks,
+    const deque<pair<PubKey, Peer>>& initDsComm, const uint64_t& index_num,
+    deque<pair<PubKey, Peer>>& newDSComm) {
+  deque<pair<PubKey, Peer>> mutable_ds_comm = initDsComm;
+
+  bool ret = true;
+
+  uint64_t prevdsblocknum =
+      m_mediator.m_dsBlockChain.GetLastBlock().GetHeader().GetBlockNum();
+  uint64_t totalIndex = index_num;
+  ShardingHash prevShardingHash =
+      m_mediator.m_dsBlockChain.GetLastBlock().GetHeader().GetShardingHash();
+
+  for (const auto& dirBlock : dirBlocks) {
+    if (typeid(DSBlock) == dirBlock.type()) {
+      const auto& dsblock = get<DSBlock>(dirBlock);
+      if (dsblock.GetHeader().GetBlockNum() != prevdsblocknum + 1) {
+        LOG_GENERAL(WARNING, "DSblocks not in sequence "
+                                 << dsblock.GetHeader().GetBlockNum() << " "
+                                 << prevdsblocknum);
+        ret = false;
+        break;
+      }
+
+      if (!CheckBlockCosignature(dsblock, mutable_ds_comm)) {
+        LOG_GENERAL(WARNING, "Co-sig verification of ds block "
+                                 << prevdsblocknum + 1 << " failed");
+        ret = false;
+        break;
+      }
```

Comment:
It's better to have a checker for the prevBlockHash also, it could be the hash of any of these 3 types of block...but seems we haven't implement the prevHash linkage for them yet when composing, let's do them together in a seperate pr.

Reply:
@kaikawaliu sure